### PR TITLE
Modify all of the iterators to allow returning an error on Next()

### PR DESCRIPTION
**Pull request by jsternberg**

### CODE REVIEW

**Comment by jsternberg** [comment]
@benbjohnson not ready to merge yet (the errors don't get propagated in some places), but this is the initial work we should build on for getting `Next()` to return an error.

**Comment by jsternberg** [comment]
@benbjohnson this should now be ready for a review and merging.

File: influxql/iterator.gen.go.tmpl [hunk 1]
```diff
@@ -298,18 +317,11 @@ func new{{$k.Name}}SortedMergeIterator(inputs []{{$k.Name}}Iterator, opt Iterato
...
-		p := input.Next()
-		if p == nil {
-			continue
-		}
-
 		// Append to the heap.
-		itr.heap = append(itr.heap, &{{$k.name}}SortedMergeHeapItem{point: p, itr: input, ascending: opt.Ascending})
+		itr.heap = append(itr.heap, &{{$k.name}}SortedMergeHeapItem{itr: input, ascending: opt.Ascending})
 	}
-	heap.Init(&itr.heap)
```
Comment (benbjohnson) [inline influxql/iterator.gen.go.tmpl]: Why is this `heap` stuff changed?

Comment (jsternberg) [inline influxql/iterator.gen.go.tmpl]: The same reason that `MergeIterator` was changed. I moved the heap initialization to the first call of `Next()`. Part of the reason for this was so that, if an error happened, it could be returned by `Next()`.

This change isn't needed as much as the one to `MergeIterator` because the wait time for a point is less, but it should still be done that way for consistency.

File: influxql/iterator.gen.go.tmpl [hunk 2]
```diff
@@ -497,33 +531,33 @@ func new{{$k.Name}}FillIterator(input {{$k.Name}}Iterator, expr Expr, opt Iterat
...
-		itr.window.name, itr.window.tags = p.Name, p.Tags
-		itr.window.time = itr.startTime
-	} else {
-		if opt.Ascending {
-			itr.window.time = itr.endTime + 1
-		} else {
-			itr.window.time = itr.endTime - 1
-		}
-	}
-	return itr
```
Comment (benbjohnson) [inline influxql/iterator.gen.go.tmpl]: Why is this removed?

File: influxql/iterator.gen.go.tmpl [hunk 3]
```diff
@@ -657,6 +691,7 @@ func (itr *{{$k.name}}InterruptIterator) Next() *{{$k.Name}}Point {
 type {{$k.name}}AuxIterator struct {
 	input      *buf{{$k.Name}}Iterator
 	output     chan *{{$k.Name}}Point
+	err        chan error
 	fields     auxIteratorFields
 	background bool
 }
```
Comment (benbjohnson) [inline influxql/iterator.gen.go.tmpl]: It seems like using a `chan struct{ *{{$k.Name}}Point, error }` would more closely match the signature of `Next()`. I'm worried that two channels could cause timing issues in the future and they may also add additional lock overhead.

File: influxql/iterator.gen.go.tmpl (hunk 2 above)
Comment (jsternberg) [inline influxql/iterator.gen.go.tmpl]: It's moved into `Next()` to avoid a call to `peek()` inside of the initializer. That way, we reduce the time that a query is uninterruptible. It also has the benefit of removing the else statement which never really worked correctly, but existed there to attempt to convince `Next()` to return nil. Now, `Next()` can just return nil instead of trying to find some weird way for these functions to communicate with each other.

File: influxql/iterator.gen.go.tmpl (hunk 3 above)
Comment (jsternberg) [inline influxql/iterator.gen.go.tmpl]: Sure. I should be able to do that. Let me try it.

Comment (jsternberg) [inline influxql/iterator.gen.go.tmpl]: Done. I also found a bug where I forgot to propagate the error to the underlying channel iterators.

**Comment by benbjohnson** [comment]
:+1:
